pkg/server/endpoints/auth.go

```diff
@@ -0,0 +1,161 @@
+package endpoints
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/spiffe/go-spiffe/v2/bundle/x509bundle"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/go-spiffe/v2/spiffetls/tlsconfig"
+	"github.com/spiffe/go-spiffe/v2/svid/x509svid"
+	"github.com/spiffe/spire/pkg/common/telemetry"
+	"github.com/spiffe/spire/pkg/server/cache/dscache"
+	"github.com/spiffe/spire/pkg/server/svid"
+	"github.com/spiffe/spire/proto/spire/common"
+)
+
+var (
+	misconfigLogMtx   sync.Mutex
+	misconfigLogTimes = make(map[string]time.Time)
+	misconfigClk      = clock.New()
+)
+
+const misconfigLogEvery = time.Minute
+
+// shouldLogFederationMisconfiguration returns true if the last time a misconfiguration
+// was logged was more than misconfigLogEvery ago.
+func shouldLogFederationMisconfiguration(tdID string) bool {
+	misconfigLogMtx.Lock()
+	defer misconfigLogMtx.Unlock()
+
+	now := misconfigClk.Now()
+	last, ok := misconfigLogTimes[tdID]
+	if !ok || now.Sub(last) >= misconfigLogEvery {
+		misconfigLogTimes[tdID] = now
+		return true
+	}
+	return false
+}
+
+// bundleGetter fetches the bundle for the given trust domain and parse it as x509 certificates.
+func (e *Endpoints) bundleGetter(ctx context.Context, td spiffeid.TrustDomain) ([]*x509.Certificate, error) {
+	commonServerBundle, err := e.DataStore.FetchBundle(dscache.WithCache(ctx), td.IDString())
+	if err != nil {
+		return nil, fmt.Errorf("get bundle from datastore: %w", err)
+	}
+	if commonServerBundle == nil {
+		if td != e.TrustDomain && shouldLogFederationMisconfiguration(td.IDString()) {
+			e.Log.
+				WithField(telemetry.TrustDomain, td.String()).
+				Warn(
+					"No bundle found for foreign admin trust domain; admins from this trust domain will not be able to connect. " +
+						"Make sure this trust domain is correctly federated.",
+				)
+		}
+		return nil, fmt.Errorf("no bundle found for trust domain %q", e.TrustDomain)
```

Comment:
Shouldn't these be logging `td`?
```suggestion
		return nil, fmt.Errorf("no bundle found for trust domain %q", td)
```

Reply:
yes!

Reply:
e70e7b0